metsim/io.py

```diff
@@ -0,0 +1,227 @@
+                calendar=params['calendar'], var_dict=params['forcing_vars'])
+            for var in params['forcing_vars'].keys():
+                met_data[var].loc[{'lat': lat, 'lon': lon}] = ds[var]
+        except (ValueError, KeyError):
+            continue
+    return met_data
+
+
+def read_ascii(data_handle, domain=None, iter_dims=['lat', 'lon'],
+               start=None, stop=None, calendar='standard',
+               var_dict=None) -> xr.Dataset:
+    """Read in an ascii forcing file"""
+    dates = date_range(start, stop, calendar=calendar)
+    names = var_dict.keys()
+    ds = pd.read_table(data_handle, header=None, delim_whitespace=True,
+                       names=names).head(len(dates))
+    ds.index = dates
+    return ds
+
+
+def read_netcdf(data_handle, domain=None, iter_dims=['lat', 'lon'],
+                start=None, stop=None, calendar='standard',
+                var_dict=None) -> xr.Dataset:
+    """Read in a NetCDF file"""
+    ds = xr.open_dataset(data_handle)
+
+    if var_dict is not None:
+        ds.rename(var_dict, inplace=True)
+
+    if start is not None and stop is not None:
+        ds = ds.sel(time=slice(start, stop))
+        dates = ds.indexes['time']
+        ds['day_of_year'] = xr.Variable(('time', ), dates.dayofyear)
+
+    if domain is not None:
+        ds.sel(**{d: domain[d] for d in iter_dims})
+    out = ds.load()
+    ds.close()
+    return out
+
+
+def read_data(data_handle, domain=None, iter_dims=['lat', 'lon'],
+              start=None, stop=None, calendar='standard',
+              var_dict=None) -> xr.Dataset:
+    """Read data directly from an xarray dataset"""
+    varlist = list(data_handle.keys())
+    if var_dict is not None:
+        data_handle.rename(var_dict, inplace=True)
+        varlist = list(var_dict.values())
+
+    if start is not None and stop is not None:
+        data_handle = data_handle[varlist].sel(time=slice(start, stop))
+        dates = data_handle.indexes['time']
+        data_handle['day_of_year'] = xr.Variable(('time', ), dates.dayofyear)
+
+    if domain is not None:
+        data_handle.sel(**{d: domain[d] for d in iter_dims})
+    out = data_handle.load()
+    data_handle.close()
+    return out
+
+
+def read_binary(data_handle, domain=None, iter_dims=['lat', 'lon'],
+                start=None, stop=None, calendar='standard',
+                var_dict=None) -> xr.Dataset:
+    """Reads a binary forcing file (VIC 4 format)"""
+    dates = date_range(start, stop, calendar=calendar)
+    n_days = len(dates)
+    type_lookup = {'signed': 'h', 'unsigned': 'H'}
+    # Pack these for nicer syntax in the loop
+    var_names = var_dict.keys()
+    data_list = [[] for var in var_dict.keys()]
+    n_vars = len(var_names)
+    params = [p for p in var_dict.values()]
+    scales = [float(s.split()[0]) for s in params]
+    datatypes = [type_lookup[s.split()[-1]] for s in params]
+    with open(data_handle, 'rb') as f:
```

Comment:
I think you can speed this by reading larger chunks (VIC ascii files are sufficiently small that we should be able to read them into memory in one go), but let's leave it for now. We can change that later.